### PR TITLE
feat: [HL-36] Create logger with default logger🍕

### PR DESCRIPTION
**Pull request by djayyab**
## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

**Review by ahmdmhasn — CHANGES_REQUESTED** [review]
Files changed should include changes related to the ticket only.

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+             line: UInt)
+}
+
+public enum LogLevel {
+    case verbose
+    case debug
+    case info
+    case warn
+    case error
+}
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]: SHould be added to a separate file

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+struct Logger {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]:
```suggestion
final class Logger {
```

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+                // NO-OP
+            }
+        }
+        if NSClassFromString("XCTest") != nil {
+            // Unit-test
+            return NoOpLogger()
+        }
+        return cachLogger
+    }()
+    static func logDebug(_ message: String,
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]:
```suggestion
    }()
    
    static func logDebug(_ message: String,
```

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                         file: StaticString = #fileID,
+                         function: StaticString = #function,
+                         line: UInt = #line) {
+        logger.log(message,
+                   level: .debug,
+                   file: file,
+                   function: function,
+                   line: line)
+    }
+    static func logInfo(_ message: String,
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]:
```suggestion
    }
    
    static func logInfo(_ message: String,
```

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                        file: StaticString = #fileID,
+                        function: StaticString = #function,
+                        line: UInt = #line) {
+        logger.log(message,
+                   level: .info,
+                   file: file,
+                   function: function,
+                   line: line)
+    }
+    static func logWarn(_ message: String,
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]:
```suggestion
    }
    
    static func logWarn(_ message: String,
```

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                        file: StaticString = #fileID,
+                        function: StaticString = #function,
+                        line: UInt = #line) {
+        logger.log(message,
+                   level: .warn,
+                   file: file,
+                   function: function,
+                   line: line)
+    }
+    static func logVerbose(_ message: String,
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]:
```suggestion
    }
    
    static func logVerbose(_ message: String,
```

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+                           file: StaticString = #fileID,
+                           function: StaticString = #function,
+                           line: UInt = #line) {
+        logger.log(message,
+                   level: .verbose,
+                   file: file,
+                   function: function,
+                   line: line)
+    }
+    static func logError(_ message: String,
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]:
```suggestion
    }
    
    static func logError(_ message: String,
```

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+                         line: UInt = #line) {
+        logger.log(message,
+                   level: .error,
+                   file: file,
+                   function: function,
+                   line: line)
+    }
+}
+
+class CachLogger: Logging {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]:
```suggestion
final class FileSystemLogger: Logging {
```

Comment (ahmdmhasn) [reply Healthy/Classes/Utilities/Logger/Logger.swift]: Also, please move this to a separate file

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+                    fileHandle.closeFile()
+                }
+                fileHandle.seekToEndOfFile()
+                fileHandle.write(data)
+            } else {
+                try data.write(to: fileURL, options: .atomic)
+            }
+        }
+    }
+}
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]:
```suggestion
// MARK: String+Helpers
private extension String {
    func appendLineToURL(fileURL: URL) throws {
        let data = self + "\n"
        try data.appendToURL(fileURL: fileURL)
    }
    
    func appendToURL(fileURL: URL) throws {
        if let data = self.data(using: String.Encoding.utf8) {
            if let fileHandle = FileHandle(forWritingAtPath: fileURL.path) {
                defer {
                    fileHandle.closeFile()
                }
                fileHandle.seekToEndOfFile()
                fileHandle.write(data)
            } else {
                try data.write(to: fileURL, options: .atomic)
            }
        }
    }
}
```

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        logger.log(message,
+                   level: .error,
+                   file: file,
+                   function: function,
+                   line: line)
+    }
+}
+
+class CachLogger: Logging {
+    //    static let shared = CachLogger()
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]:
```suggestion
```

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    private let logFileName = "app.log"
+    private let logQueue = DispatchQueue(label: "", qos: .background, attributes: .concurrent)
+    func log(_ message: String, level: LogLevel, file: StaticString, function: StaticString, line: UInt) {
+        let timestamp = DateFormatter.localizedString(from: Date(), dateStyle: .medium, timeStyle: .medium)
+        let logMessage = "\(timestamp) - \(level) - \(file) - \(function) - \(line)"
+        logCache.append(logMessage)
+        logQueue.async {  [weak self] in self?.flushLogCache()}
+        print(logMessage)
+    }
+    private func flushLogCache() {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]:
```suggestion
    private let logQueue = DispatchQueue(label: "", qos: .background, attributes: .concurrent)
    
    func log(_ message: String, level: LogLevel, file: StaticString, function: StaticString, line: UInt) {
        let timestamp = DateFormatter.localizedString(from: Date(), dateStyle: .medium, timeStyle: .medium)
        let logMessage = "\(timestamp) - \(level) - \(file) - \(function) - \(line)"
        logCache.append(logMessage)
        logQueue.async {  [weak self] in self?.flushLogCache()}
        print(logMessage)
    }
    
    private func flushLogCache() {
```

File: Healthy/Classes/AppDelegate.swift [hunk 12]
```diff
@@ -7,6 +7,10 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     func application(_ application: UIApplication,
                      didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         // Override point for customization after application launch.
+
+        _ = CachLogger()
+        Logger.logWarn("⚠️")
+        Logger.logError("❌")
```
Comment (ahmdmhasn) [inline Healthy/Classes/AppDelegate.swift]:
```suggestion

```

File: Healthy/Resources/Assets.xcassets/images/background-splash.imageset/Rectangle 6.pdf
Comment (ahmdmhasn) [inline Healthy/Resources/Assets.xcassets/images/background-splash.imageset/Rectangle 6.pdf:1]: Should be removed

File: Healthy/Resources/Assets.xcassets/images/background-splash.imageset/Contents.json
Comment (ahmdmhasn) [inline Healthy/Resources/Assets.xcassets/images/background-splash.imageset/Contents.json:1]: Should be removed

File: Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift:1]: Irrelevant changes. Should be reverted.

File: Healthy/Classes/Utilities/Logger/String+Helpers.swift [hunk 13]
```diff
@@ -0,0 +1,20 @@
+import Foundation
+extension String {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/String+Helpers.swift]: This is tightly related to the logger. Let's add them as private extension in the same logger file.

File: Healthy/Classes/Utilities/Logger/Logging.swift [hunk 14]
```diff
@@ -0,0 +1,8 @@
+import Foundation
+public protocol Logging {
+    func log(_ message: String,
+             level: LogLevel,
+             file: StaticString,
+             function: StaticString,
+             line: UInt)
+}
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logging.swift]:
```suggestion
import Foundation

/// The Logging protocol defines a standardized way of logging messages in an application or system.
public protocol Logging {
    func log(_ message: String,
             level: LogLevel,
             file: StaticString,
             function: StaticString,
             line: UInt)
}
```

File: Healthy/Classes/Generated/UIImage.Generated.swift
Comment (ahmdmhasn) [inline Healthy/Classes/Generated/UIImage.Generated.swift:1]: Changes should be reverted

File: Healthy/Classes/AppDelegate.swift [hunk 15]
```diff
@@ -3,9 +3,14 @@ import UIKit
...
-    func application(_ application: UIApplication,
-                     didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+    func application(_ application: UIApplication, didFinishLaunchingWithOptions
+            launchOptions:
+            [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         // Override point for customization after application launch.
+
+        _ = FileSystemLogger()
+        Logger.logWarn("⚠️")
+        Logger.logError("❌")
```
Comment (ahmdmhasn) [inline Healthy/Classes/AppDelegate.swift]: Dead code, should be removed and added to the pull request description or code documentation as an example

File: Healthy/Classes/Utilities/Logger/FileSystemLogger.swift [hunk 16]
```diff
@@ -0,0 +1,38 @@
+import Foundation
+final class FileSystemLogger: Logging {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/FileSystemLogger.swift]:
```suggestion
import Foundation

final class FileSystemLogger: Logging {
```

Please keep a message after import statement and between methods!

File: Healthy/Classes/Utilities/Logger/FileSystemLogger.swift [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+            .first?.appendingPathComponent(logFileName) else {
+            print("Error getting log URL")
+            return
+        }
+        do {
+            let logString = logCache.joined(separator: "\n")
+            try logString.appendLineToURL(fileURL: logURL)
+            logCache.removeAll()
+        } catch {
+            print("Error writing log to file:\(error)")
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/FileSystemLogger.swift]:
```suggestion
            assertionFailure("Error writing log to file:\(error)")
```

File: Healthy/Classes/Utilities/Logger/FileSystemLogger.swift [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+                                                      dateStyle: .medium,
+                                                      timeStyle: .medium)
+        let logMessage = "\(timestamp) - \(level) - \(file) - \(function) - \(line)"
+        logCache.append(logMessage)
+        logQueue.async {  [weak self] in self?.flushLogCache()}
+        print(logMessage)
+    }
+    
+    
+    private func flushLogCache() {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/FileSystemLogger.swift]:
```suggestion
    private let logQueue = DispatchQueue(label: "", qos: .background, attributes: .concurrent)
    
    func log(_ message: String,
             level: LogLevel,
             file: StaticString,
             function: StaticString,
             line: UInt) {
        let timestamp = DateFormatter.localizedString(from: Date(),
                                                      dateStyle: .medium,
                                                      timeStyle: .medium)
        let logMessage = "\(timestamp) - \(level) - \(file) - \(function) - \(line)"
        logCache.append(logMessage)
        logQueue.async {  [weak self] in self?.flushLogCache()}
        print(logMessage)
    }
    
    private func flushLogCache() {
```

File: Healthy/Classes/Utilities/Logger/FileSystemLogger.swift [hunk 19]
```diff
@@ -0,0 +1,38 @@
+import Foundation
+final class FileSystemLogger: Logging {
+    private var logCache: [String] = []
+    private let logFileName = "app.log"
+    private let logQueue = DispatchQueue(label: "", qos: .background, attributes: .concurrent)
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/FileSystemLogger.swift]:
```suggestion
    private let logQueue = DispatchQueue(label: "com.motoon.default-logger", qos: .background, attributes: .concurrent)
```

Dispatch queue label should never be empty.

File: Healthy/Classes/Utilities/Logger/LogLevel.swift [hunk 20]
```diff
@@ -0,0 +1,8 @@
+import Foundation
+public enum LogLevel {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/LogLevel.swift]:
```suggestion
public enum LogLevel {
```

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+                     function: StaticString,
+                     line: UInt) {
+                // NO-OP
+            }
+        }
+        if NSClassFromString("XCTest") != nil {
+            // Unit-test
+            return NoOpLogger()
+        }
+        return cachLogger
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]:
```suggestion
        }
        
        if NSClassFromString("XCTest") != nil {
            // Unit-test
            return NoOpLogger()
        }
        
        return cachLogger
```

File: HealthyTests/HealthyTests.swift [hunk 22]
```diff
@@ -6,6 +6,7 @@ final class HealthyTests: XCTestCase {
     override func setUpWithError() throws {
         // Put setup code here. This method is called before the invocation of
         // each test method in the class.
+        Logger.logWarn("Warning from Unit Test")
```
Comment (ahmdmhasn) [inline HealthyTests/HealthyTests.swift]:
```suggestion
```

File: Healthy.xcodeproj/project.pbxproj [hunk 23]
```diff
@@ -23,11 +23,21 @@
 		6DF5250C2A1E855F0027502C /* LoginViewModelType.swift in Sources */ = {isa = PBXBuildFile; fileRef = 6DF5250B2A1E855F0027502C /* LoginViewModelType.swift */; };
 		79B65D282A0FB4310042CC46 /* UIButton+Style.swift in Sources */ = {isa = PBXBuildFile; fileRef = 79B65D272A0FB4310042CC46 /* UIButton+Style.swift */; };
 		B22218482A10C08D0036B9A8 /* UILabel+Style.swift in Sources */ = {isa = PBXBuildFile; fileRef = B22218472A10C08D0036B9A8 /* UILabel+Style.swift */; };
+		B8574F042A2022AA0028FC4B /* HealthyValidator .swift in Sources */ = {isa = PBXBuildFile; fileRef = B8574F032A2022AA0028FC4B /* HealthyValidator .swift */; };
```
Comment (ahmdmhasn) [inline Healthy.xcodeproj/project.pbxproj]: There're alot of irrelevant changes that should be reverted.

File: Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift
Comment (djayyab) [reply Healthy/Classes/Utilities/ValidatorProtocol/HealthyValidationRule .swift:1]: How can I revert the changes? manually? these changes make conflict with my branch so I add them

File: Healthy/Classes/Utilities/Logger/Logger.swift [hunk 24]
```diff
@@ -0,0 +1,75 @@
+import Foundation
+final class Logger {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/Logger.swift]:
```suggestion
import Foundation

final class Logger {
```

File: Healthy/Classes/Utilities/Logger/LogLevel.swift (hunk 20 above)
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/LogLevel.swift]:
```suggestion
import Foundation

public enum LogLevel {
```

File: Healthy/Classes/Utilities/Logger/FileSystemLogger.swift [hunk 25]
```diff
@@ -0,0 +1,34 @@
+import Foundation
+final class FileSystemLogger: Logging {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/FileSystemLogger.swift]:
```suggestion
import Foundation

final class FileSystemLogger: Logging {
```

File: Healthy/Classes/Utilities/Logger/FileSystemLogger.swift [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+             line: UInt) {
+        let timestamp = DateFormatter.localizedString(from: Date(),
+                                                      dateStyle: .medium,
+                                                      timeStyle: .medium)
+        let logMessage = "\(timestamp) - \(level) - \(file) - \(function) - \(line)"
+        logCache.append(logMessage)
+        logQueue.async {  [weak self] in self?.flushLogCache()}
+        print(logMessage)
+    }
+    private func flushLogCache() {
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/FileSystemLogger.swift]:
```suggestion
    }
    
    private func flushLogCache() {
```

File: Healthy/Classes/Utilities/Logger/FileSystemLogger.swift [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+        guard let logURL = FileManager.default
+            .urls(for: .cachesDirectory, in: .userDomainMask)
+            .first?.appendingPathComponent(logFileName) else {
+            print("Error getting log URL")
+            return
+        }
```
Comment (ahmdmhasn) [inline Healthy/Classes/Utilities/Logger/FileSystemLogger.swift:29]:
```suggestion
        }
        
```

**Review by ahmdmhasn — APPROVED** [review]

**Pull request merged**